src/digraph.rs

```diff
@@ -2183,6 +2195,79 @@ impl PyDiGraph {
+    /// value from the matrix.
+    ///
+    /// This differs from the
+    /// :meth:`~retworkx.PyDiGraph.from_adjacency_matrix` in that the type of
+    /// the elements of the input matrix in this method must be a ``complex``
+    /// (specifically a ``numpy.complex128``) and the output graph edge weights
+    /// will be ``complex`` too. While in
+    /// :meth:`~retworkx.PyDiGraph.from_adjacency_matrix` the matrix elements
+    /// are of type ``float`` (specifically ``numpy.float64``) and the edge
+    /// weights in the output graph will be ``float`` too.
+    ///
+    /// :param ndarray matrix: The input numpy array adjacency matrix to create
+    ///     a new :class:`~retworkx.PyDiGraph` object from. It must be a 2
+    ///     dimensional array and be a ``complex``/``np.complex128`` data type.
+    /// :param float null_value: An optional float that will treated as a null
+    ///     value. If any element in the input matrix is this value it will be
+    ///     treated as not an edge. By default this is ``0.0``
+    ///
+    /// :returns: A new graph object generated from the adjacency matrix
+    /// :rtype: PyDiGraph
+    #[staticmethod]
+    #[pyo3(text_signature = "(matrix, /)")]
+    pub fn from_complex_adjacency_matrix<'p>(
+        py: Python<'p>,
+        matrix: PyReadonlyArray2<'p, Complex64>,
+        null_value: Option<Complex64>,
+    ) -> PyDiGraph {
+        let null: Complex64 = null_value.unwrap_or_else(Complex64::zero);
+        let array = matrix.as_array();
+        let shape = array.shape();
+        let mut out_graph = StableDiGraph::<PyObject, PyObject>::new();
+        let _node_indices: Vec<NodeIndex> = (0..shape[0])
+            .map(|node| out_graph.add_node(node.to_object(py)))
+            .collect();
+        array
+            .axis_iter(Axis(0))
+            .enumerate()
+            .for_each(|(index, row)| {
+                let source_index = NodeIndex::new(index);
+                for target_index in 0..row.len() {
+                    if null.is_nan() {
+                        if !row[[target_index]].is_nan() {
+                            out_graph.add_edge(
+                                source_index,
+                                NodeIndex::new(target_index),
+                                row[[target_index]].to_object(py),
+                            );
+                        }
+                    } else if row[[target_index]] != null {
+                        out_graph.add_edge(
+                            source_index,
+                            NodeIndex::new(target_index),
+                            row[[target_index]].to_object(py),
+                        );
+                    }
+                }
+            });
+
+        PyDiGraph {
+            graph: out_graph,
+            cycle_state: algo::DfsSpace::default(),
+            check_cycle: false,
+            node_removed: false,
+            multigraph: true,
+        }
```

Comment:
I think it is worth trying to avoid the `from_adjacency_matrix` code duplication. Maybe we can replace the logic with a generic function for [`numpy::TypeNum`](https://pyo3.github.io/rust-numpy/numpy/types/trait.TypeNum.html) and call that instead?

Reply:
I was working in that direction before pushing up this WIP because I was blocked on the num-complex version issue. The thing I was getting most stuck on besides the num-complex versioning issues was the handling of the trait for `nan` handling. I agree we should do this as generically as possible instead of duplicating the code when we circle back to this after the next pyo3 release

Reply:
Does the [`num_traits::FloatCore`](https://docs.rs/num-traits/0.2.14/num_traits/float/trait.FloatCore.html) trait help you with handling `nan`? it is implemented for both float and complex types, and guarantees that `is_nan` will be present

Reply:
I believe `num_traits::FloatCore` is not implemented for complex.

We can avoid the code duplication either with a macro  (so we dont have to define any traits that a type must implement) or use a generic function over a type that implements the appropriate traits (`Copy`, `num_traits::Zero`, `std::cmp::PartialEq`, `numpy::Element`, `pyo3::ToPyObject`) plus an extra custom trait like:

```rust
trait IsNan {
    fn is_nan(self) -> bool;
}

/// https://doc.rust-lang.org/nightly/src/core/num/f64.rs.html#441
impl IsNan for f64 {
    #[inline]
    fn is_nan(self) -> bool {
        self != self
    }
}

/// https://docs.rs/num-complex/0.4.0/src/num_complex/lib.rs.html#572-574
impl IsNan for Complex64 {
    #[inline]
    fn is_nan(self) -> bool {
        self.re.is_nan() || self.im.is_nan()
    }
}
```

Reply:
Done in: https://github.com/Qiskit/retworkx/pull/412/commits/c4460837e6d080d1c0dbd6191d519886fe96e66c